contracts/NativeTokenWallet.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "./interfaces/IGatewayStructs.sol";
+import "./interfaces/INativeTokenWallet.sol";
+import "./NativeTokenPredicate.sol";
+
+/**
+    @title NativeToken
+    @notice Native token contract
+ */
+// solhint-disable reason-string
+contract NativeTokenWallet is
+    Initializable,
+    OwnableUpgradeable,
+    UUPSUpgradeable,
+    IGatewayStructs,
+    INativeTokenWallet
+{
+    address public predicate;
+    uint256 public totalSupply;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize() public initializer {
+        __Ownable_init(msg.sender);
+        __UUPSUpgradeable_init();
+    }
+
+    function _authorizeUpgrade(
+        address newImplementation
+    ) internal override onlyOwner {}
+
+    function setDependencies(
+        address _predicate,
+        uint256 _tokenSupply
+    ) external onlyOwner {
+        if (_predicate == address(0)) revert ZeroAddress();
+        predicate = _predicate;
+        totalSupply = _tokenSupply;
+    }
+
+    /**
+     * @notice Deposits an amount of tokens to a particular address
+     * @dev Can only be called by the predicate or owner address
+     * @param _account Account of the user to mint the tokens to
+     * @param _amount Amount of tokens to mint to the account
+     * @return bool Returns true if function call is successful
+     */
+    function deposit(
+        address _account,
+        uint256 _amount
+    ) external onlyPredicateOrOwner returns (bool) {
+        if (_account == address(0)) revert ZeroAddress();
```

Comment:
i am not sure if we should just skip if receiver is zero address. Otherwise, whole batch will be reverted. I am not sure if this check exists on oracle